src/mbgl/style/style.cpp

```diff
@@ -158,7 +158,30 @@ void Style::addSource(std::unique_ptr<Source> source) {
     sources.emplace_back(std::move(source));
 }
 
+struct SourceIdUsageEvaluator {
+    const std::string& sourceId;
+
+    bool operator()(BackgroundLayer&) { return false; }
+    bool operator()(CustomLayer&) { return false; }
+
+    template <class LayerType>
+    bool operator()(LayerType& layer) {
+        return layer.getSourceID() == sourceId;
+    }
+};
+
 std::unique_ptr<Source> Style::removeSource(const std::string& id) {
+    // Check if source is in use
+    SourceIdUsageEvaluator sourceIdEvaluator {id};
+    auto layerIt = std::find_if(layers.begin(), layers.end(), [&](const auto& layer) {
+        return layer->accept(sourceIdEvaluator);
+    });
+
+    if (layerIt != layers.end()) {
+        Log::Warning(Event::General, "Source '%s' is in use, cannot remove", id.c_str());
+        return nullptr;
```

Comment:
@jfirebaugh Yes, a `nullptr` was already returned in case the source doesn't exist [here](https://github.com/mapbox/mapbox-gl-native/blob/9f7aad7e0c29f5ed5399b993b2be8d0d83a3db91/src/mbgl/style/style.cpp#L190). I've added tests for the [Android](https://github.com/mapbox/mapbox-gl-native/pull/9129/commits/978d60ba801fd3a5514f8c2111135dc391f16b93) and [iOS/macos](https://github.com/mapbox/mapbox-gl-native/pull/9129/commits/9f7aad7e0c29f5ed5399b993b2be8d0d83a3db91) SDKs to ensure this.